### DISCUSSION TITLE
Naver sign-in does not adhere to OAuth 2.0 specs

### DISCUSSION THREAD

**Discussion by susemeee** [nextauthjs/next-auth#9313] [Help]
While implementing `NaverProvider` with auth.js (next-auth) v5, the flow causes OperationProcessingError denoting invalid response type of `expires_in` field.

```
[auth][cause]: OperationProcessingError: "response" body "expires_in" property must be a positive number
```

Official docs of Naver states the value should be integer, but actual value is coming with string typed.

<img width="912" alt="Screenshot 2023-12-05 at 12 00 18 PM" src="https://github.com/nextauthjs/next-auth/assets/4835442/d71ebfa0-6e8c-4428-bcba-b23e2e1f9d3b">

<img width="244" alt="Screenshot 2023-12-05 at 12 04 17 PM" src="https://github.com/nextauthjs/next-auth/assets/4835442/5dafc342-14ae-46ca-b76b-f3ac4e063fa1">

and the `expires_in` value should be typed as number as per RFC 6749 page 43.

> The parameters are included in the entity-body of the HTTP response
   using the "application/json" media type as defined by [[RFC4627](https://www.rfc-editor.org/rfc/rfc4627)].  The
   parameters are serialized into a JavaScript Object Notation (JSON)
   structure by adding each parameter at the highest structure level.
   Parameter names and string values are included as JSON strings.
   Numerical values are included as JSON numbers.  The order of
   parameters does not matter and can vary.

This will cause inability implementing `NaverProvider` with v5.

I think I can't do anything meaningful right now besides reporting it to Naver but leaving this for further triaging.

**Comment by susemeee** [comment]
- [Report to Naver](https://developers.naver.com/forum/posts/35298)

**Comment by kms0219kms** [comment]
Naver’s official answer for this issue(English below):

안녕하세요 수세미님
네이버 로그인 입니다

먼저 불편을 드려 죄송합니다.

OAuth2.0 스펙과 다르게 제공되고 있으나, 사이드 이펙트의 우려로 이미 제공되고 있는 스펙을 수정하는 것은 어렵다는 점 양해 부탁드립니다

문서 상으로 OAuth2.0 스펙과 다르게 동작된다는걸 명시하고, 가능하면 next-auth 에도 해당 사항이 반영될 수 있도록 하겠습니다

감사합니다

(English)
Hello, this is Naver login.

First of all, we apologize for the inconvenience.

Although it is provided differently from the OAuth2.0 specification, please understand that it is difficult to modify the already provided specification due to concerns about side-effect.

In the document, we will specify that it operates differently from the OAuth2.0 specification, and if possible, we will ensure that this is reflected in next-auth as well.

Thanks

**Comment by kms0219kms** [comment]
Due to Naver's response, they will not fix the issue now. Therefore, we need to hard-code only for this Naver's case.

**Reply by teamsmiley** [reply]
어떻게 하드코딩을 하면되나요?

**Reply by kms0219kms** [reply]
@teamsmiley 저는 직접 네이버로그인을 할 수 있도록, next-auth를 다 뜯어서 찾아 보았고,
https://github.com/nextauthjs/next-auth/issues/8868#issuecomment-1807229914 이 방법과 굉장히 유사하게 커스텀하여 사용했습니다.

다만 현재로서는 https://github.com/nextauthjs/next-auth/discussions/9313#discussioncomment-9342069 이 방법이 더 좋은것 같네요.

**Comment by curlydoodle** [comment]
I'm getting the same error with Azure AD

**Comment by lodado** [comment]
how to fix it manually ?

**Comment by jacobhjkim** [comment]
@teamsmiley @curlydoodle @lodado 

I patched the underlying dependency library `oauth4webapi`. You can override your dependency in your `package.json` like below.
```json
"overrides": {
  "oauth4webapi": "npm:@jacobkim/oauth4webapi@^2.10.4"
}
```
This will fix the issue with Azure AD B2C and Naver provider. Here is a detailed explanation [https://jacob.kim/en/articles/next-auth-expires-in](https://jacob.kim/en/articles/next-auth-expires-in).

---

(Korean)

저와 같이 Auth.js (next-auth) v5와 네이버 로그인을 사용하시려고 하시는 분을 위해 해결 방안을 블로그에 써두었습니다.
[https://jacob.kim/ko/articles/next-auth-expires-in](https://jacob.kim/ko/articles/next-auth-expires-in)

요약하자면, 
```json
"overrides": {
  "oauth4webapi": "npm:@jacobkim/oauth4webapi@^2.10.4"
}
```
package.json에 `oauth4webapi`만 override 해주면 됩니다.

**Reply by kms0219kms** [reply]
@jacobhjkim 님, 보다 쉬운 해결방법을 찾아주셔서 너무 좋습니다.
다만 포크하신게 업데이트가 안되고 있다보니, dependency outdated 문제도 있고 하여 저는 계속 Custom Provider 방식을 사용하고 있는데요.
주기적으로 Fork를 Sync하여 업데이트 해 오실 의향은 없으신지 여쭙고 싶습니다.

**Reply by jacobhjkim** [reply]
@kms0219kms updated the fork to `v2.17.0`!
